File: instrumentation/newrelic-scala-cats-api/src/main/java/com/newrelic/cats/api/TraceOps_Instrumentation.java
package com.newrelic.cats.api;

import cats.effect.IO;
import com.newrelic.api.agent.weaver.Weave;
import com.newrelic.api.agent.weaver.Weaver;

@Weave(originalName = "com.newrelic.cats.api.TraceOps$")
public class TraceOps_Instrumentation {
  public <S> IO<S> txn(final IO<S> body) {
    return Util.wrapTrace(Weaver.callOriginal());
  }
}

File: instrumentation/newrelic-scala-cats-api/src/main/scala/com/newrelic/cats/api/TraceOps_Instrumentation.scala
package com.newrelic.cats.api

import cats.effect.Sync
import com.newrelic.api.agent.weaver.Weaver
import com.newrelic.api.agent.weaver.scala.{ScalaMatchType, ScalaWeave}

@ScalaWeave(`type` = ScalaMatchType.Object, `originalName` = "com.newrelic.cats.api.TraceOps")
class TraceOps_Instrumentation {
  def txn[S, F[_]:Sync](body: F[S]): F[S] = {
    Util.wrapTrace(Weaver.callOriginal)
  }
}
